HW2/P3/mandelbrot.pyx

```diff
@@ -3,45 +3,99 @@ cimport numpy as np
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
-
-    with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+    for i in range(in_coords.shape[0]):
+        for j in range(in_coords.shape[1]):
+            in_coords_real[i,j]=in_coords[i,j].real
+            in_coords_imag[i,j]=in_coords[i,j].imag
+    with nogil, parallel(num_threads=4):
+        #setup thread local buffer
+        out_counts_temp = <float *> malloc(sizeof(float) * 8)
+        for i in prange(in_coords.shape[0],schedule='static',chunksize=1):           
+            for j in range(in_coords.shape[1]/8):
+                c_real = AVX.make_float8(in_coords_real[i,j*8+7],
+                                    in_coords_real[i,j*8+6],
+                                    in_coords_real[i,j*8+5],
+                                    in_coords_real[i,j*8+4],
+                                    in_coords_real[i,j*8+3],
+                                    in_coords_real[i,j*8+2],
+                                    in_coords_real[i,j*8+1],
+                                    in_coords_real[i,j*8+0])
+                c_imag = AVX.make_float8(in_coords_imag[i,j*8+7],
+                                    in_coords_imag[i,j*8+6],
+                                    in_coords_imag[i,j*8+5],
+                                    in_coords_imag[i,j*8+4],
+                                    in_coords_imag[i,j*8+3],
+                                    in_coords_imag[i,j*8+2],
+                                    in_coords_imag[i,j*8+1],
+                                    in_coords_imag[i,j*8+0])
+                z_real = AVX.float_to_float8(0.0)
+                z_imag = AVX.float_to_float8(0.0)
+                avx_out_counts=AVX.float_to_float8(0.0)
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    # #do the magnitude squared computation in two steps...
+                    magnitude_squared=AVX.mul(z_real,z_real)
+                    magnitude_squared=AVX.fmadd(z_imag,z_imag,magnitude_squared)
+
+                    #mag^2<4, return 1. 
+                    cont_case=AVX.less_than_equal(magnitude_squared,AVX.float_to_float8(4))
+
+                    #if mag^2<4, cont_case will be -NaN, otherwise 0
+                    #When all iterations for all 8 elements are done, 
+                    #all bits of cont_case will be 0.
+                    if AVX.signs(cont_case) == 0:
                         break
-                    z = z * z + c
-                out_counts[i, j] = iter
 
+                    z_real_new = AVX.add(AVX.sub(AVX.mul(z_real,z_real),AVX.mul(z_imag,z_imag)),c_real)
```

Comment:
You could achieve more speedup by removing some redundant multiplies of z_imag. Once when calculating magnitude_squared and once when calculating z_real_new. Same for z_real.